hazelcast/src/hazelcast/client/compact.cpp

```diff
@@ -0,0 +1,678 @@
+ */
+int64_t
+fingerprint64(const std::string& type_name,
+              std::map<std::string, field_descriptor>& fields)
+{
+    uint64_t fingerPrint = fingerprint64(INIT, type_name);
+    fingerPrint = fingerprint64(fingerPrint, (int)fields.size());
+    for (const auto& entry : fields) {
+        const field_descriptor& descriptor = entry.second;
+        fingerPrint = fingerprint64(fingerPrint, entry.first);
+        fingerPrint = fingerprint64(fingerPrint, (int)descriptor.field_kind);
+    }
+    return static_cast<int64_t>(fingerPrint);
+}
+
+} // namespace rabin_finger_print
+
+bool
+kind_size_comparator(const field_descriptor* i, const field_descriptor* j)
+{
+    auto i_kind_size =
+      field_operations::get(i->field_kind).kind_size_in_byte_func();
+    auto j_kind_size =
+      field_operations::get(j->field_kind).kind_size_in_byte_func();
+    return i_kind_size < j_kind_size;
+}
+
+schema::schema(
+  std::string type_name,
+  std::unordered_map<std::string, field_descriptor>&& field_definition_map)
+  : type_name_(std::move(type_name))
+  , field_definition_map_(std::move(field_definition_map))
+{
+    std::vector<field_descriptor*> fixed_size_fields;
+    std::vector<field_descriptor*> boolean_fields;
+    std::vector<field_descriptor*> variable_size_fields;
+
+    std::map<std::string, field_descriptor> sorted_fields(
+      field_definition_map_.begin(), field_definition_map_.end());
+    for (auto& item : sorted_fields) {
+        field_descriptor& descriptor = item.second;
+        field_kind kind = descriptor.field_kind;
+        if (field_operations::get(kind).kind_size_in_byte_func() ==
+            field_kind_based_operations::VARIABLE_SIZE) {
+            variable_size_fields.push_back(&descriptor);
+        } else if (kind == field_kind::BOOLEAN) {
+            boolean_fields.push_back(&descriptor);
+        } else {
+            fixed_size_fields.push_back(&descriptor);
+        }
+    }
+
+    std::sort(
+      fixed_size_fields.begin(), fixed_size_fields.end(), kind_size_comparator);
+
+    int offset = 0;
+    for (auto descriptor : fixed_size_fields) {
+        descriptor->offset = offset;
+        offset += field_operations::get(descriptor->field_kind)
+                    .kind_size_in_byte_func();
+    }
+
+    int bit_offset = 0;
+    for (auto descriptor : boolean_fields) {
+        descriptor->offset = offset;
+        descriptor->bit_offset = bit_offset % util::Bits::BITS_IN_BYTE;
+        bit_offset++;
+        if (bit_offset % util::Bits::BITS_IN_BYTE == 0) {
+            offset += 1;
+        }
+    }
+    if (bit_offset % util::Bits::BITS_IN_BYTE != 0) {
+        offset += 1;
+    }
+
+    fixed_size_fields_length_ = offset;
+
+    int index = 0;
+    for (auto descriptor : variable_size_fields) {
+        descriptor->index = index;
+        index++;
+    }
+
+    for (auto& item : sorted_fields) {
+        field_definition_map_[item.first] = item.second;
+    }
```

Comment:
Do we need this part? The Key->Value mappings of both `field_definition_map_` and `sorted_fields` should be the same. The only difference is that `sorted_fields` has these mappings sorted according to the key.  So, it is useful during the schema id generation. 

However, why do you set the already existing mappings to the `field_definition_map_` again?

Reply:
field_definition_map_ is the one stored on schema and will be used later. 
sorted_fields is a temporary copy of field_definition_map_. We first populated the index,offset info on the field_descriptors exist on sorted_fields. Since it is a copy, they are not reflected to `field_definition_map_` . So I had to copy it back.

Reply:
@sancar  Not critical but you could have used emplace and std::move to eliminate copy.